File: src/main/java/com/clinica/atendimento/dto/ConvenioDTO.java
package com.clinica.atendimento.dto;

import com.clinica.atendimento.model.Convenio;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ConvenioDTO {
    private Long id;
    private String nome;

    // Static method to convert from entity to DTO
    public static ConvenioDTO fromEntity(Convenio convenio) {
        if (convenio == null) {
            return null;
        }

        return ConvenioDTO.builder()
                .id(convenio.id())
                .nome(convenio.nome())
                .build();
    }

    // Method to convert from DTO to entity
    public Convenio toEntity() {
        // Using the constructor for backward compatibility
        // return new Convenio(id, nome);

        // Using the builder pattern
        Convenio.ConvenioBuilder builder = Convenio.builder();

        // Only set ID if it's not null
        if (id != null) {
            builder.id(id);
        }

        return builder
                .nome(nome)
                .build();
    }
}

File: src/main/kotlin/com/clinica/atendimento/dto/ConvenioDTO.kt
package com.clinica.atendimento.dto

import com.clinica.atendimento.model.Convenio
import com.fasterxml.jackson.annotation.JsonProperty

data class ConvenioDTO(
    @JsonProperty
    var id: Long? = null,

    @JsonProperty
    var nome: String? = null
) {
    // Fluent accessors to maintain compatibility with Java code
    fun id(id: Long?): ConvenioDTO {
        this.id = id
        return this
    }

    fun nome(nome: String?): ConvenioDTO {
        this.nome = nome
        return this
    }

    // Method to convert from DTO to entity
    fun toEntity(): Convenio {
        val builder = Convenio.builder()

        // Only set ID if it's not null
        id?.let { builder.id(it) }

        return builder
            .nome(nome)
            .build()
    }

    companion object {
        // Static method to convert from entity to DTO
        @JvmStatic
        fun fromEntity(convenio: Convenio?): ConvenioDTO? {
            if (convenio == null) {
                return null
            }

            return ConvenioDTO.builder()
                .id(convenio.id())
                .nome(convenio.nome())
                .build()
        }

        @JvmStatic
        fun builder(): ConvenioDTOBuilder {
            return ConvenioDTOBuilder()
        }
    }

    class ConvenioDTOBuilder {
        private val instance = ConvenioDTO()

        fun id(id: Long?): ConvenioDTOBuilder {
            instance.id = id
            return this
        }

        fun nome(nome: String?): ConvenioDTOBuilder {
            instance.nome = nome
            return this
        }

        fun build(): ConvenioDTO {
            return instance
        }
    }
}